### ISSUE TITLE
Unix HttpClient throwing an exception I can't catch

### ISSUE THREAD

**Issue by i8beef** [dotnet/corefx#30805]
I'm running an application with the Microsoft 2.1 Docker images. The app is using the Microsoft.Extensions.Hosting generic host stuff to implement a long running poll that calls a POST endpoint via a static HttpClient. Now I have an unfortunate issue where the device I'm calling returns a non-standard result, which on Windows, would give me a "WinHttpException: The server returned an invalid or unrecognized response”. I have exhausted my options to make it do anything else, the device in question simply does not follow protocol. So my HttpClient.SendAsync() call is wrapped in a specific catch to try and swallow this:

```
            try
            {
                var response = await SendAsyncWithTimeout(requestMessage, cancellationToken)
                    .ConfigureAwait(false);
            }
            catch (HttpRequestException ex) when (ex.InnerException != null && ex.InnerException.Message == "The server returned an invalid or unrecognized response")
            {
                // This will ALWAYS throw an exception because the Apex API doesn't even return a valid HTTP protocol response
                // Thus it is ignored. This is probably one of the more evil things I've ever written...
            }
```

When this app is run on Windows (At least in a debugger), it swallows the error as expected and things work fine.

But I found when I run it on a Linux instance, it fails to catch the exception:

```
Unhandled Exception: System.Net.Http.HttpRequestException: An error occurred while sending the request. ---> System.IO.IOException: The server returned an invalid or unrecognized response.
   at System.Net.Http.HttpConnection.FillAsync()
   at System.Net.Http.HttpConnection.ReadNextResponseHeaderLineAsync(Boolean foldedHeadersAllowed)
   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
... My code here ...
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.ThreadPoolWorkQueue.Dispatch()

```

Anyone see anything obvious here I'm missing, or something thats different in the Linux implementation that would cause this?

**Comment by davidsh** [comment]
>Now I have an unfortunate issue where the device I'm calling returns a non-standard result,

Can you provide a Fiddler or WireShark trace?  We'd like to see exactly what the "non-standard" server response is.  That will enable us to diagnose to see if it is something we can "ignore" safely without throwing an exception.

**Comment by i8beef** [comment]
I found it. 

The device in question returns a 0-length response, which would result in a cUrl error #52. I don't have a capture.

The ISSUE is the fact that I need to swallow this and I'm keying on the InnerException.Message. It appears on Windows, the WinHttpException has a message of "The server returned an invalid or unrecognized response"... but in the Core implementation it'll throw an exception with the message ""The server returned an invalid or unrecognized response." Notice that period.

Once I changed the catch clause to look for StartsWith("The server returned an invalid or unrecognized response") instead of == "The server returned an invalid or unrecognized response", the exception is caught on both platforms. I don't know if one should be changed to match, or if this is an incredibly minor "breaking" change, but that was the issue.
